File: 100-SameTree/100-SameTree.js
// Last updated: 4/29/2026, 10:34:46 AM
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {boolean}
 */
var isSameTree = function(p, q) {
    if(!p && !q) {
        return true;
    } else if(!p || !q) {
        return false;
    }
    
    const compare = (n1, n2) => {
        if((!n1 && n2) || (n1 && !n2)) {
            return false;
        }
        if(n1.val !== n2.val) {
            return false;
        }
        if(n1.left || n2.left) {
            if(!compare(n1.left, n2.left)) {
                return false;
            }
        }
        if(n1.right || n2.right) {
            if(!compare(n1.right, n2.right)) {
                return false;
            }
        }
        return true;
    }
    
    return compare(p, q);
};

File: 100-SameTree/100-SameTree.py
# Last updated: 4/29/2026, 10:34:51 AM
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        def compare(p, q):
            if p == None and q == None:
                return True
            elif p != None and q == None:
                return False
            elif p == None and q != None:
                return False
            if p.val != q.val:
                return False
            leftRes = compare(p.left, q.left)
            if not leftRes:
                return False
            
            rightRes = compare(p.right, q.right)
            if not rightRes:
                return False
            
            return True
        return compare(p, q)